docs/design/discovery-service.md

```diff
@@ -0,0 +1,119 @@
+
+* Only one of the EnvoyConfigRevisions holds the current version of the config. This is called the **published version** and is marked in the EnvoyConfigRevision with the `RevisionPublished` condition. It is the EnvoyConfig controller the one deciding which of its owned EnvoyConfigRevisions is the one actually published. The algorithm used to decide which is one it should be is:
+
+    1. EnvoyConfig controller keeps a list of EnvoyConfigRevision references in `status.configRevisions`, ordered by time of publication. The last published revision holds the highest array index position.
+    2. Revisions are versioned by computing the hash of the `spec.envoyResources` field.
+    3. The revision with the highest array index that is not marked with the `RevisionTainted` condition is marked with the `RevisionPublished` condition, effectively getting it published.
+    4. All other owned EnvoyConfigRevisions get the `RevisionPublished` condition set to `false`.
+
+* When an EnvoyConfig resource gets updated, the hash of `spec.envoyResources` is recalculated and a new EnvoyConfigRevision for that hash is created. If an EnvoyConfigRevision already exists that matches the hash, the existing reference in `status.configRevisions` that points to that EnvoyConfigRevision gets moved to the array's highest index position, effectively triggering the publication of that revision.
+
+* The EnvoyConfigRevision controller watches events on EnvoyConfigRevision custom resources. Whenever it receives an event on one, it checks if the revision is marked as published. If so, loads the envoy resources from serizalized format into proto message objects and writes them to the xDS server in-memory cache. The xDS server will start delivering the new config to the envoy proxies as soon as it detects changes in the in-memory cache.
+
+* The xDS server detects when the config sent to an envoy proxy is not valid due to the [NACKs](https://www.envoyproxy.io/docs/envoy/v1.16.0/api-docs/xds_protocol#basic-protocol-overview) defined in the xDS protocol. This is done by a callback function that inspects the DiscoveryRequest messages received by the server looking for NACKs. Whenever a NACK is detected, the callback function marks the relevant EnvoyConfigRevision custom resource with the `RevisionTainted` condition. This triggers a rollback process and the last not tainted revision in the list will get published instead. The EnvoyConfig custom resource will get the `Rollback` status in the `status.CacheState` field. If there is not a single revision untainted in the EnvoyConfig's revision list, the EnvoyConfig will set the `RollbackFailed` status in the `status.CacheState` field and the failing config will be still published until the config gets fixed by the user and a new publication process is triggered.
+
+The following image depicts the described process.
+
+![Discovery service](discovery-service.svg)
```

Comment:
It might be useful to further clarify where in the diagram is the XDs Server, making it more explicit for people not as familiar with envoy

Reply:
I also see a webhook server component in the server but no references of it in the text.
Would it be worth it to explain that part in this document or you don't see a need for that?

Reply:
Just a nitpick, I see some icons where it says `crd` but I guess it would be `cr`. The message is understood though so whatever you prefer

Reply:
> It might be useful to further clarify where in the diagram is the XDs Server located, making it more explicit for people not as familiar with envoy

I'll update the diagram to make it more clear, thnx!

> I also see a webhook server component in the server but no references of it in the text.
> Would it be worth it to explain that part in this document, reference another doc or you don't see a need for that?

The webhook part is documented elsewhere, document being added in this other PR #51 

> Just a nitpick, I see some icons where it says `crd` but I guess it would be `cr`. The message is understood though so whatever you prefer

Totally true but there is no CR icon in lucidcharts :joy:

Reply:
:+1: